### DISCUSSION TITLE
Introducing a lockfile to Glasskube

### DISCUSSION THREAD

**Discussion by jakepage91** [glasskube/glasskube#1097] [Ideas]
During the last [community call](https://youtu.be/sx_qWIUZwO4?si=ePjAi1zD1R3znD1g) (min 34:07) the question of `lockfiles` being used was brought up. As a way to pin Glasskube docker images to specific lockfile hashes for example. 

Opening this discussion here to make sure this is included in future conversation about future Glasskube feature conversations. Feel free to add to the conversation.

**Comment by linkdd** [comment]
So, what I meant regarding Docker images was this:

It's DevOps 101 to never use the `latest` tag in production and to always pin your dependencies. But did you know that other tags like `vX.Y.Z` are not immutable as well? Nothing prevents the upstream from replacing the image with another one. And if your Pod use `imagePullPolicy: Always`, congratulations: you are vulnerable to supply chain attacks! :tada:

The only thing immutable is the actual Docker image digest (which is assigned by the registry when pushing the image, it's a SHA256 hash).

I generally use the tool [kbld](https://carvel.dev/kbld/) in production to resolve the image tags to an actual digest and really pin the dependencies.

The goal of a lockfile is to maintain the list of pinned dependencies and their hashes. To me, it includes:

- Helm chart versions for packages that wraps an Helm chart, and actual hash of the helm chart
- Package dependencies versions
- Pod's docker image versions
- Probably other stuff I'm forgetting

The whole point is to make supply chain attacks more difficult, and then use tools like Renovate / Dependabot / whatever to manage dependency upgrades.
